crates/contracts/solidity/Solver.sol

```diff
@@ -104,4 +105,26 @@ contract Solver {
             _simulationOverhead += gasStart - gasleft() + 4460;
         }
     }
+
+    /// @dev Helper function that reads and stores the balances of the `settlementContract` for each token in `tokens`.
+    /// @param tokens - list of tokens used in the trade
+    /// @param settlementContract - the settlement contract whose balances are being read
+    function _storeSettlementBalances(address[] calldata tokens, ISettlement settlementContract) internal {
+        for (uint256 i = 0; i < tokens.length; i++) {
+            this.storeBalance(tokens[i], address(settlementContract), false);
+        }
+    }
+
+    /// @dev Executes the settlement and measures the gas used.
+    /// @param settlementContract The address of the settlement contract.
+    /// @param settlementCall The calldata for the settlement function.
+    /// @return gasUsed The amount of gas used during the settlement execution.
+    function _executeSettlement(
+        address settlementContract,
+        bytes calldata settlementCall
+    ) private returns (uint256 gasUsed) {
+        uint256 gasStart = gasleft();
+        address(settlementContract).doCall(settlementCall);
+        gasUsed = gasStart - gasleft() - _simulationOverhead;
+    }
```

Comment:
Extracting some code into separate functions is required since the EVM's limit is 16 stack slots. After updating the  `swap` function, compilation started to fail: 
```
Stack too deep. Try compiling with `--via-ir` (cli) or the equivalent `viaIR: true` (standard JSON) while enabling the optimizer. Otherwise, try removing local variables.
```
While the proposed IR optimizer [is not experimental](https://docs.soliditylang.org/en/latest/ir-breaking-changes.html) anymore, it still has implications that might change some of the semantics. So, instead of using it, I decided that minor refactoring is safer.